contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,672 @@
+        require(_sharesAmount > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = validators[validator];
+        uint256 _bnbAmount = IStakePool(valInfo.poolModule).undelegate(delegator, _sharesAmount);
+
+        emit Undelegated(validator, delegator, _bnbAmount);
+
+        _updateEligibleValidators(UpdateDirection.Down, validator, valInfo.voteAddress);
+    }
+
+    function redelegate(address srcValidator, address dstValidator, uint256 _sharesAmount) public onlyInitialized validatorExist(srcValidator) validatorExist(dstValidator) validatorNotJailed(dstValidator) whenNotPaused {
+        require(_sharesAmount > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory srcValInfo = validators[srcValidator];
+        Validator memory dstValInfo = validators[dstValidator];
+
+        uint256 _bnbAmount = IStakePool(srcValInfo.poolModule).redelegate(delegator, _sharesAmount);
+        IStakePool(dstValInfo.poolModule).delegate{value: _bnbAmount}(delegator);
+
+        emit Redelegated(srcValidator, dstValidator, delegator, _bnbAmount);
+
+        bytes memory srcValVoteAddr = srcValInfo.voteAddress;
+        bytes memory dstValVoteAddr = dstValInfo.voteAddress;
+        _updateEligibleValidators(UpdateDirection.Down, srcValidator, srcValVoteAddr);
+        _updateEligibleValidators(UpdateDirection.Up, dstValidator, dstValVoteAddr);
+    }
+
+    function claimSecurityDeposit(uint256 _sharesAmount) external onlyInitialized validatorExist(msg.sender) {
+        Validator memory valInfo = validators[msg.sender];
+        address pool = valInfo.poolModule;
+        require(IStakePool(pool).balanceOf(address(this)) >= _sharesAmount, "NOT_ENOUGH_SHARES");
+
+        emit SecurityDepositClaimed(msg.sender, _sharesAmount);
+
+        IStakePool(pool).transfer(msg.sender, _sharesAmount);
+        if (IStakePool(pool).getSecurityDepositBNB() < minSelfDelegationBNB) {
+            valInfo.jailed = true;
+            emit ValidatorJailed(msg.sender);
+        }
+    }
+
+    function claim(address validator, uint256 requestNumber) external onlyInitialized validatorExist(validator) {
+        uint256 _bnbAmount = IStakePool(validators[validator].poolModule).claim(msg.sender, requestNumber);
+
+        emit Claimed(validator, msg.sender, _bnbAmount);
+    }
+
+    /*----------------- system functions -----------------*/
+    function distributeReward(address validator) external payable onlyInitialized onlyValidatorContract validatorNotJailed(validator) {
+        IStakePool(validators[validator].poolModule).distributeReward{value: msg.value}();
+    }
+
+    function getEligibleValidators() external view returns (IBSCValidatorSet.Validator[] memory, bytes[] memory) {
+        return (eligibleValidators, eligibleValidatorVoteAddrs);
+    }
+
+    function downtimeSlash(address validator, uint256 height) external onlyInitialized onlySlash {
```

Comment:
why `height` is required, can't we get it from context?

Reply:
`SlashIndicator` will get it from context and pass it to here. I add this to be more flexible